server/src/main/java/com/tencent/rss/server/buffer/ShuffleBuffer.java

```diff
@@ -0,0 +1,374 @@
+    long eventId = ShuffleFlushManager.ATOMIC_EVENT_ID.getAndIncrement();
+    final ShuffleDataFlushEvent event = new ShuffleDataFlushEvent(
+        eventId,
+        appId,
+        shuffleId,
+        startPartition,
+        endPartition,
+        size,
+        spBlocks,
+        isValid,
+        this);
+    inFlushBlockMap.put(eventId, spBlocks);
+    blocks.clear();
+    updateShuffleSize(shuffleSizeMap, appId, shuffleId, size);
+    size = 0;
+    return event;
+  }
+
+  // update metadata of shuffleId -> size
+  private void updateShuffleSize(
+      Map<String, Map<Integer, AtomicLong>> shuffleSizeMap,
+      String appId,
+      Integer shuffleId,
+      long flushSize) {
+    Map<Integer, AtomicLong> shuffleIdToSize = shuffleSizeMap.get(appId);
+    if (shuffleIdToSize != null) {
+      AtomicLong atomicLong = shuffleIdToSize.get(shuffleId);
+      if (atomicLong != null) {
+        atomicLong.addAndGet(-flushSize);
+      }
+    }
+  }
+
+  public List<ShufflePartitionedBlock> getBlocks() {
+    return blocks;
+  }
+
+  public long getSize() {
+    return size;
+  }
+
+  public boolean isFull() {
+    return size > capacity;
+  }
+
+  public synchronized void clearInFlushBuffer(long eventId) {
+    inFlushBlockMap.remove(eventId);
+  }
+
+  @VisibleForTesting
+  public Map<Long, List<ShufflePartitionedBlock>> getInFlushBlockMap() {
+    return inFlushBlockMap;
+  }
+
+  // 1. generate buffer segments and other info: if blockId exist, start with which eventId
+  // 2. according to info from step 1, generate data
+  public synchronized ShuffleDataResult getShuffleData(
+      long lastBlockId, int readBufferSize) {
+    try {
+      List<BufferSegment> bufferSegments = Lists.newArrayList();
+      ActualReadInfo actualReadInfo = updateBufferSegments(
+          lastBlockId, readBufferSize, bufferSegments);
+      if (actualReadInfo.actualReadSize > 0) {
+        byte[] data = new byte[actualReadInfo.actualReadSize];
+        updateShuffleData(actualReadInfo.lastBlockId, readBufferSize,
+            data, actualReadInfo.fromEventId);
+        return new ShuffleDataResult(data, bufferSegments);
+      }
+    } catch (Exception e) {
+      LOG.error("Exception happened when getShuffleData in buffer", e);
+    }
+    return new ShuffleDataResult();
+  }
+
+  // here is the rule to read data in memory:
+  // 1. read from inFlushBlockMap order by eventId asc, then from blocks
+  // 2. if can't find lastBlockId, means related data may be flushed to storage, repeat step 1
+  private ActualReadInfo updateBufferSegments(
+      long lastBlockId,
+      long readBufferSize,
+      List<BufferSegment> bufferSegments) throws Exception {
+    long fromEventId = INVALID_EVENT_ID;
+    long currentBlockId = lastBlockId;
+    List<Long> sortedEventId = sortFlushingEventId();
+    int offset = 0;
+    int eventIndex = 0;
+    long updatedLastBlockId = lastBlockId;
+    boolean hasBlockId = false;
```

Comment:
If block is flushed,  we need to read from the start of buffer. Is possible to read duplicated data.

Reply:
yes, it will cause a little performance problem, but it's ok for the first step

Reply:
At least, we should left some `to do` or comments.

Reply:
ok

Reply:
And log some necessary information.